### PR TITLE
Run Functional Tests in CI

### PR DESCRIPTION
**Pull request by SharonHart**
Add additional job to CI pipeline for factional tests validation that:
1. Starts presidio cluster with docker-compose
2. Bootstraps the functional tests venv  
3. Runs the tests with pytest

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -89,4 +89,34 @@ jobs:
...
+            action: Run services
+            dockerComposeFile: docker-compose.yml
+            buildImages: true
+
+        - task: UsePythonVersion@0
+          inputs:
+            versionSpec: '3.8'
+
+        - script: |
+            python -m venv env
```
Comment (balteravishay) [inline azure-pipelines.yml:110]: Additional "set -eux" to bash tasks for logging and errors

File: azure-pipelines.yml [hunk 2]
```diff
@@ -89,4 +89,34 @@ jobs:
...
+            source env/bin/activate
+            python -m pip install --upgrade pip
+            pip install -r requirements.txt
+            pip install pytest-azurepipelines
+          workingDirectory: functional-tests
+          displayName: Install requirements
+
+        - script: |
+            source env/bin/activate
+            pytest
```
Comment (balteravishay) [inline azure-pipelines.yml:120]: Does this also output test results to ado?

**Review by balteravishay — COMMENTED** [review]
Some comments

File: azure-pipelines.yml (hunk 1 above)
Comment (SharonHart) [reply azure-pipelines.yml:110]: Thanks missed that. exiting with nonzero code is not failing ado?

File: azure-pipelines.yml (hunk 2 above)
Comment (SharonHart) [reply azure-pipelines.yml:120]: Yes, checked rhat

File: azure-pipelines.yml (hunk 1 above)
Comment (SharonHart) [reply azure-pipelines.yml:111]: Done

**Review by navalev — CHANGES_REQUESTED** [review]
looks good. Lets move the functional tests to a seperate template to make sure the pipeline is readable

**Comment by SharonHart** [comment]
@navalev Will do, good idea

**Comment by SharonHart** [comment]
@navalev Done, approve once we merge the base and I'll rebase

**Review by navalev — APPROVED** [review]

File: presidio-anonymizer/Dockerfile [hunk 3]
```diff
@@ -10,5 +10,5 @@ RUN pipenv sync
 
 COPY . /usr/bin/${NAME}/
 
-EXPOSE ${FLASK_RUN_PORT}
-CMD pipenv run flask run --host 0.0.0.0
\ No newline at end of file
+EXPOSE ${PORT}
+CMD pipenv run python app.py --host 0.0.0.0
```
Comment (navalev) [inline presidio-anonymizer/Dockerfile:14]: since we are going to change this to flask run, please add a TODO comment with the relevant user id/bug

**Review by navalev — CHANGES_REQUESTED** [review]

File: presidio-anonymizer/Dockerfile (hunk 3 above)
Comment (SharonHart) [reply presidio-anonymizer/Dockerfile]: Done

File: presidio-anonymizer/Dockerfile [hunk 4]
```diff
@@ -10,5 +10,6 @@ RUN pipenv sync
 
 COPY . /usr/bin/${NAME}/
 
-EXPOSE ${FLASK_RUN_PORT}
-CMD pipenv run flask run --host 0.0.0.0
\ No newline at end of file
+# [ADO-2680] Use FLASK_RUN_PORT and flask run --host 0.0.0.0
```
Comment (shiranr) [inline presidio-anonymizer/Dockerfile:13]: A really small remark:
# [TODO ADO-2680] Use FLASK_RUN_PORT and flask run --host 0.0.0.0
Please add the TODO - this way it is being recognized as a TODO and marked accordingly in the code.

Comment (SharonHart) [reply presidio-anonymizer/Dockerfile:13]: missed that

File: .pipelines/templates/functional-tests.yml [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+    - script: |
+          set -eux  # fail on error
+          python -m venv env
+          source env/bin/activate
+          python -m pip install --upgrade pip
+          pip install -r requirements.txt
+          pip install pytest-azurepipelines
+      workingDirectory: functional-tests
+      displayName: Install requirements
```
Comment (shiranr) [inline .pipelines/templates/functional-tests.yml:21]: required dependencies maybe?

**Review by shiranr — APPROVED** [review]

**Review by shiranr — APPROVED** [review]

**Review by balteravishay — APPROVED** [review]

**Pull request merged**